query-algebrizer/src/validate.rs

```diff
@@ -74,6 +75,46 @@ pub fn validate_or_join(or_join: &OrJoin) -> Result<()> {
     }
 }
 
+pub fn validate_not_join(not_join: &NotJoin) -> Result<()> {
+    // Grab our mentioned variables and ensure that the rules are followed.
+    match not_join.unify_vars {
+        UnifyVars::Implicit => {
+            if not_join.clauses.len() < 2 {
+                Ok(())
+            } else {
+                let mut clauses = not_join.clauses.iter();
+                let template = clauses.next().unwrap().collect_mentioned_variables();
+                for clause in clauses {
+                    if template != clause.collect_mentioned_variables() {
+                        bail!(ErrorKind::NonMatchingVariablesInNotClause);
```

Comment:
This whole clause can go.

A `not` has only a single body, made up of multiple clauses; it's the negated version of an implied `and`. Simply being `Implicit` satisfies intrinsic validation.

(The clause might be _extrinsically_ invalid if its unify list isn't already bound outside the pattern, but you can't answer that question here.)